### PR TITLE
SCC modularization

### PR DESCRIPTION
**Pull request by awnawab**
I've split the SCC transform into six transforms:

- `SCCBaseTransformation`
- `SCCDevectorTransformation`
- `SCCDemoteTransformation`
- `SCCRevectorTransformation`
- `SCCHoistTransformation`
- `SCCAnnotateTransformation`

The original `SingleColumnCoalescedTransformation` also still exists, but it's basically now a wrapper script that calls the methods contained in the above transformations.

I've updated `loki-transform.py` to run each one of the six above one-by-one rather than running the unified SCC transform. This is regression tested on the ecWam physics against the original SCC transform and produces string-identical results.

I still have to refactor the tests to better test the new modular transforms individually, but before I do that I would really appreciate your input on whether I am the right path.

NB: this is stacked on top of a small bug fix in #89.

### CODE REVIEW

**Comment by awnawab** [comment]
Regression tested against cloudsc and ecWam physics. I've left the old `SingleColumnCoalescedTransformation` (including some tests that call it) in. It's now basically a wrapper script that calls the methods in the modular transformations. In principle this can, and perhaps should, be removed.

File: transformations/transformations/scc_annotate.py
Comment (mlange05) [inline transformations/transformations/scc_annotate.py:166]: [No action] Indeed, this mechanism isn't great, but will likely only become redundant once we can preempt repeated processing, which should come with the scheduler rewrites. Ok for now, I htink.

**Review by mlange05 — COMMENTED** [review]
Ok, first of all, this is looking really great, and in addition is exposing all the right dependencies to make start working on Transformation composition. The test coverage is also very, very nice, although we might be able to reduce the replication at a later stage once composition can be done neatly. All in all, I'm super happy with this, so, many thanks!!!

I'm marking this as "Comments" for now, until I've run the full ec-phys regression, as I've left a few minor comments about the location of some utilities. None of these are showstoppers though, and I'll approve once regression test has run (ongoing).

File: transformations/transformations/scc_annotate.py
Comment (reuterbal) [reply transformations/transformations/scc_annotate.py:166]: The scheduler already ensures that each item is visited only once, so shouldn't this instead be replaced by the typical pattern?
```python
if (item := kwargs.get('item')) and item.name != routine.name.lower():
    return
```
I think this should become part of the dispatch in the `Transformation.apply` routine eventually.

**Review by mlange05 — APPROVED** [review]
Ok, EC-physics regression test is clean and comments are not considered showstoppers, so I'll green-light this now. I'll wait for secondary review and/or responses, but ready from my side.

File: transformations/transformations/scc_demote.py
Comment (reuterbal) [inline transformations/transformations/scc_demote.py:87]: Same here

File: transformations/transformations/scc_base.py
Comment (reuterbal) [inline transformations/transformations/scc_base.py:206]: Same here

File: transformations/transformations/scc_devector.py
Comment (reuterbal) [inline transformations/transformations/scc_devector.py:132]: Same here

File: transformations/transformations/scc_hoist.py
Comment (reuterbal) [inline transformations/transformations/scc_hoist.py:184]: Same here

File: transformations/transformations/scc_revector.py
Comment (reuterbal) [inline transformations/transformations/scc_revector.py:72]: Same here

**Review by reuterbal — APPROVED** [review]
This is great and well-tested! I'm a bit concerned about the large number of modules, which isn't necessarily the Pythonic way (where I see module files more like directories than individual source files in other languages) but that's a matter of taste.

My only real request would be to remove the `self._processed` checks and instead use the established pattern of checking against the item name, which we use elswhere and which achieves the same.

For things flagged in the comments and that are not being addressed now, it might make sense to create issues to not forget them down the line.

File: transformations/transformations/scc_annotate.py
Comment (mlange05) [reply transformations/transformations/scc_annotate.py:166]: Yes, I agree. I was planning on moving these bail-out mechanism when I switch to composed SCC for full ec-physics, but we can bring them in line here already

Comment (reuterbal) [reply transformations/transformations/scc_annotate.py:166]: I think it's not necessary to move this already to the base class as part of this PR, but would rather not introduce a second pattern.

File: transformations/transformations/scc_annotate.py
Comment (awnawab) [reply transformations/transformations/scc_annotate.py]: Yes agreed. I'm not sure how this ended up in the diff, but I think a subsequent commit already moved this to the base.

File: loki/transform/transform_array_indexing.py
Comment (awnawab) [reply loki/transform/transform_array_indexing.py:35]: Yes agreed, will move it there and add tests. Thanks!

File: transformations/transformations/scc_annotate.py
Comment (awnawab) [reply transformations/transformations/scc_annotate.py:166]: Yes, fully agreed. I should have realised that this is just duplicating an existing mechanism. I'll replace all instances of the `self._processed` checks with B's suggestion above.

File: transformations/transformations/scc_revector.py
Comment (awnawab) [reply transformations/transformations/scc_revector.py:72]: Yup, will replace all of these with the existing mechanism. Thanks again!

File: transformations/tests/test_single_column_coalesced.py
Comment (awnawab) [reply transformations/tests/test_single_column_coalesced.py:23]: Done, thanks!

File: scripts/loki_transform.py
Comment (awnawab) [reply scripts/loki_transform.py:41]: I've put `SingleColumnCoalescedTransformation` into a separate module `single_column_coalesced_wrapper.py` to prevent a circular import. Since this transformation is earmarked for deletion in the near future anyway, also makes sense to put it in a separate module.

**Comment by awnawab** [comment]
Thanks a lot for your comments! I've implemented all of them, please let me know what you think.

**Comment by reuterbal** [comment]
Thanks so much, looks great! Going in...

**Pull request merged**
